File: pdam-backend/pdam.api/src/main/java/com/salesianostriana/dam/pdam/api/user/dto/GetUserDto.java
package com.salesianostriana.dam.pdam.api.user.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.salesianostriana.dam.pdam.api.user.model.User;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

import java.time.LocalDateTime;
import java.util.Objects;
import java.util.UUID;

@AllArgsConstructor
@NoArgsConstructor
@Data
@SuperBuilder
public class GetUserDto {

    protected UUID id;
    protected String userName;
    protected String fullName;
    protected String imgPath;
    protected int followers;
    protected int countOfPosts;
    protected boolean verified;
    protected boolean followedByUser;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd/MM/yyyy HH:mm:ss")
    protected LocalDateTime createdAt;

    public static GetUserDto of(User c){
        return GetUserDto.builder()
                .id(c.getId())
                .userName(c.getUsername())
                .fullName(c.getFullName())
                .imgPath(c.getImgPath())
                .followers(c.getFollowers() == null ? 0 : c.getFollowers().size())
                .countOfPosts(c.getPublishedPosts() == null ? 0 : c.getPublishedPosts().size())
                .verified(c.isVerified())
                .createdAt(c.getCreatedAt())
                .build();
    }

    public static GetUserDto ofs(User c, User loggedUser){
        return GetUserDto.builder()
                .id(c.getId())
                .userName(c.getUsername())
                .fullName(c.getFullName())
                .imgPath(c.getImgPath())
                .followers(c.getFollowers() == null ? 0 : c.getFollowers().size())
                .countOfPosts(c.getPublishedPosts() == null ? 0 : c.getPublishedPosts().size())
                .verified(c.isVerified())
                .followedByUser(c.getFollowers().stream().filter(u -> Objects.equals(u.getId(), loggedUser.getId())).toList().size() > 0)
                .createdAt(c.getCreatedAt())
                .build();
    }

}

File: pdam-frontend/pdam_app/lib/models/user/GetUserDto.dart
class GetUserDto {
  late String id;
  late String userName;
  late String fullName;
  late String imgPath;
  late int followers;
  late int countOfPosts;
  late bool verified;
  late bool followedByUser;
  late String createdAt;

  GetUserDto(
      {required this.id,
      required this.userName,
      required this.fullName,
      required this.imgPath,
      required this.followers,
      required this.countOfPosts,
      required this.verified,
      required this.followedByUser,
      required this.createdAt});

  GetUserDto.fromRegisterReqest(GetUserDto getUserDto) {
    this.id = getUserDto.id;
    this.userName = getUserDto.userName;
    this.fullName = getUserDto.fullName;
    this.imgPath = getUserDto.imgPath;
    this.followers = getUserDto.followers;
    this.countOfPosts = getUserDto.countOfPosts;
    this.verified = getUserDto.verified;
    this.followedByUser = getUserDto.followedByUser;
    this.createdAt = getUserDto.createdAt;
  }

  GetUserDto.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    userName = json['userName'];
    fullName = json['fullName'];
    imgPath = json['imgPath'];
    followers = json['followers'];
    countOfPosts = json['countOfPosts'];
    verified = json['verified'];
    followedByUser = json['followedByUser'];
    createdAt = json['createdAt'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this.id;
    data['userName'] = this.userName;
    data['fullName'] = this.fullName;
    data['imgPath'] = this.imgPath;
    data['followers'] = this.followers;
    data['countOfPosts'] = this.countOfPosts;
    data['verified'] = this.verified;
    data['followedByUser'] = this.followedByUser;
    data['createdAt'] = this.createdAt;
    return data;
  }
}
